Network Topology Dashboard Update

Dhruv-J opened the pull request:
This PR updates the Network Topology dashboard to include visualizations for
multiple networking layers. Specifically, the following has been changed:

1. The source code of DependencyPanel has been modified to include the
DiagramController class, which acts as a rendering controller for the
mermaid diagram. This is needed after an update to the mermaid library,
which changes the functionality of the existing mermaidAPI render function.

2. The @grafana-toolkit tool will be deprecated soon, and will be replaced
by @grafana/create-plugin. With the replacement, the directory structure of
the plugin has been significantly changed, with a .config folder containing
the specific configuration files for the plugin's build and run, with general
configuration files in the base directory used to extend the default configs.

3. A unit test for the DependencyPanel has been added, and it generates a
simple graph and checks the DOM for the expected graph string being present.

Additionally, this PR builds upon #417, which adds additional fields for L7
visibility.

Dhruv-J commented:
<img width="2308" alt="Screenshot 2023-09-27 at 3 00 53 AM" src="https://github.com/antrea-io/theia/assets/10016630/7dfb4072-f64f-4d1e-b4b9-4f81eb7a51fa">

Currently, the legend is shown as a paragraph, but I can make it more visual as well.

Dhruv-J commented:
For reviewers, a lot of files needed to be added as a result of the yarn tool being updated from toolkit to the grafana command itself. This update comes with a lot more surrounding configuration, and is needed for anyone wanting to build the plugin locally. The contents of the .config/ folder can be ignored, as they were automatically generated by the command yarn create @grafana/plugin migrate. The additional configuration I needed to get it working was the file webpack.config.ts file. To sum up, the code outside of the plugin directory, and both DiagramController.tsx and DependencyPanel.tsx are the main files within the PR that need to be reviewed.

yuntanghsu commented:
Why do we have duplicated connections in layer 4/7 Topology? Or are they two different connections?
For Layer 4 Topology, is svc_antrea-test-3 the endpoint of web-server-3?

Dhruv-J commented:
> Why do we have duplicated connections in layer 4/7 Topology? Or are they two different connections?
They are two different connections
> For Layer 4 Topology, is svc_antrea-test-3 the endpoint of web-server-3?
Yes, the goal of the layer 4 topology is to show destination Pod and Service, I'll update the legend to reflect that.

heanlan commented:
Please correct me if I'm wrong. I remember last time we decided to use the chord diagrams to show the flows and use a table to display all the L7 fields. Did I miss anything?

Dhruv-J commented:
> Please correct me if I'm wrong. I remember last time we decided to use the chord diagrams to show the flows and use a table to display all the L7 fields. Did I miss anything?

Not wrong at all, I'm splitting them up into two PRs, with the latter having the second dashboard with the chord diagram and httpvals table. Since httpvals contains nested JSON, some massaging to the data is required.

yuntanghsu commented on build/charts/theia/README.md:
```diff
@@ -60,7 +60,7 @@ Kubernetes: `>= 1.16.0-0`
 | grafana.homeDashboard | string | `"homepage.json"` | Default home dashboard. |
 | grafana.image | object | `{"pullPolicy":"IfNotPresent","repository":"projects.registry.vmware.com/antrea/theia-grafana","tag":"8.3.3"}` | Container image used by Grafana. |
-| grafana.installPlugins | list | `["https://downloads.antrea.io/artifacts/grafana-custom-plugins/theia-grafana-sankey-plugin-1.0.2.zip;theia-grafana-sankey-plugin","https://downloads.antrea.io/artifacts/grafana-custom-plugins/theia-grafana-chord-plugin-1.0.1.zip;theia-grafana-chord-plugin","https://downloads.antrea.io/artifacts/grafana-custom-plugins/theia-grafana-dependency-plugin-1.0.2.zip;the ...
+| grafana.installPlugins | list | `["https://downloads.antrea.io/artifacts/grafana-custom-plugins/theia-grafana-sankey-plugin-1.0.2.zip;theia-grafana-sankey-plugin","https://downloads.antrea.io/artifacts/grafana-custom-plugins/theia-grafana-chord-plugin-1.0.1.zip;theia-grafana-chord-plugin","https://github.com/Dhruv-J/grafana-dependency-plugin/archive/refs/tags/v0.zip;theia-grafana-dependency-plu ...
```
Are we able to change the domain name to `downloads.antrea.io`?

Dhruv-J replied:
yeah, this is just for testing sake, if anyone would like to run locally, they don't need additional config. upon approval of the PR, i'll adjust the link back to the proper format, since I can build and upload the plugin one final time

yuntanghsu commented on build/charts/theia/provisioning/dashboards/network_topology_dashboard.json:
```diff
@@ -32,15 +32,56 @@
+            }
+          },
+          "queryType": "sql",
+          "rawSql": "SELECT sourcePodName, sourcePodLabels, sourcePodNamespace, sourceNodeName, destinationPodName, destinationPodLabels, destinationNodeName, destinationServicePortName, octetDeltaCount FROM flows\nWHERE sourcePodNamespace NOT IN ('kube-system', 'flow-visibility', 'flow-aggregator')\nAND destinationPodNamespace NOT IN ('kube-system', 'flow-visibility', 'flow-aggregator')\nAND ( d ...
```
Please correct me if I'm wrong. 
I think we don't have case where `destinationPodName == '' and sourcePodName == ''`.?

Dhruv-J replied:
I had it as a failsafe, but I'll remove that conditional

tushartathgur replied:
Same as below

yuntanghsu commented on build/charts/theia/provisioning/dashboards/network_topology_dashboard.json:
```diff
@@ -32,15 +32,56 @@
+        "type": "grafana-clickhouse-datasource",
+        "uid": "PDEE91DDB90597936"
+      },
+      "description": "The different colors of the lines mean different codes, but are labelled by the content length of the flow.\nGreen: Success",
```
Missing description for other colors?

Dhruv-J replied:
I'll remove description, as the legend describes the colors anyways.

yuntanghsu commented on build/charts/theia/provisioning/dashboards/network_topology_dashboard.json:
```diff
@@ -102,6 +143,6 @@
+  "version": 3,
   "weekStart": ""
-}
+}
```
Missing empty line at the end of the file.

yuntanghsu commented on build/charts/theia/provisioning/datasources/migrators/000005_0-6-0.down.sql:
```diff
@@ -212,3 +212,9 @@ ALTER TABLE tadetector_local
     DROP COLUMN aggType,
     DROP COLUMN direction,
     DROP COLUMN podName;
+ALTER TABLE flows
```
Should these changes be in 000005_0-6-0.down.sql or in next version like 000006_0-7-0.down.sql?

Dhruv-J replied:
I think these are changes part of Tushar's commit and PR, I just have them in this PR as well.

tushartathgur replied:
PR has been updated, Please use the new fields and files

yuntanghsu commented on build/charts/theia/provisioning/datasources/migrators/000005_0-6-0.up.sql:
```diff
@@ -151,3 +151,9 @@ ALTER TABLE tadetector_local
     ADD COLUMN aggType String,
     ADD COLUMN direction String,
     ADD COLUMN podName String;
+ALTER TABLE flows
```
ditto

yuntanghsu commented on build/charts/theia/values.yaml:
```diff
@@ -211,7 +211,7 @@ grafana:
     - https://downloads.antrea.io/artifacts/grafana-custom-plugins/theia-grafana-sankey-plugin-1.0.2.zip;theia-grafana-sankey-plugin
     - https://downloads.antrea.io/artifacts/grafana-custom-plugins/theia-grafana-chord-plugin-1.0.1.zip;theia-grafana-chord-plugin
-    - https://downloads.antrea.io/artifacts/grafana-custom-plugins/theia-grafana-dependency-plugin-1.0.2.zip;theia-grafana-dependency-plugin
+    - https://github.com/Dhruv-J/grafana-dependency-plugin/archive/refs/tags/v0.zip;theia-grafana-dependency-plugin
```
ditto

yuntanghsu commented on plugins/grafana-custom-plugins/grafana-dependency-plugin/.config/Dockerfile:
```diff
@@ -0,0 +1,16 @@
+
+# Inject livereload script into grafana index.html
+USER root
+RUN sed -i 's/<\/body><\/html>/<script src=\"http:\/\/localhost:35729\/livereload.js\"><\/script><\/body><\/html>/g' /usr/share/grafana/public/views/index.html
```
Missing empty line here.

yuntanghsu commented on plugins/grafana-custom-plugins/grafana-dependency-plugin/.eslintrc:
```diff
@@ -0,0 +1,3 @@
+{
+  "extends": "./.config/.eslintrc"
+}
```
ditto

yuntanghsu commented on plugins/grafana-custom-plugins/grafana-dependency-plugin/.prettierrc.js:
```diff
@@ -1,3 +1,4 @@
-};
+  // Prettier configuration provided by Grafana scaffolding
+  ...require("./.config/.prettierrc.js")
+};
```
ditto

yuntanghsu reviewed (commented):
Some comments:
1. Do we need a .config folder? I don't see this folder under other plugins?
2. Do you think we can add these changes to the e2e test or UT?
3. Can you update network-flow-visibility.md to reflect these changes?

Dhruv-J commented:
> 1. Do we need a .config folder? I don't see this folder under other plugins?

Yes, as I mentioned in my comment earlier, the .config folder is created as a result of upgrading the tool used to create, build, and sign plugins. Older plugins don't have this since they haven't been rebuilt or upgraded to have it, but it is a necessary folder for anyone wishing to build the plugin locally or make changes to its source code. The additional config file I added was weback.config.ts within the base grafana-dependency-plugin directory, which was provided by a Grafana dev.

> 2. Do you think we can add these changes to the e2e test or UT?

The network topology dashboard already has a unit test, which I included when I originally made the dashboard. It simply checks the query to see if it returns expected values.

> 3. Can you update network-flow-visibility.md to reflect these changes?

Yes, will do.

tushartathgur commented on build/yamls/flow-visibility.yml:
```diff
@@ -3571,11 +3626,11 @@ data:
               },
               "queryType": "sql",
-              "rawSql": "SELECT sourcePodName, sourcePodLabels, sourcePodNamespace, sourceNodeName, destinationPodName, destinationPodLabels, destinationNodeName, destinationServicePortName, octetDeltaCount FROM flows\nWHERE sourcePodNamespace NOT IN ('kube-system', 'flow-visibility', 'flow-aggregator')\nAND destinationPodNamespace NOT IN ('kube-system', 'flow-visibility', 'flow-aggregator')\nAND ...
+              "rawSql": "SELECT sourcePodName, sourcePodLabels, sourcePodNamespace, sourceNodeName, sourceIP, sourceTransportPort, destinationPodName, destinationPodLabels, destinationNodeName, destinationServicePortName, destinationIP, destinationTransportPort, octetDeltaCount, httpVals FROM flows\nWHERE sourcePodNamespace NOT IN ('kube-system', 'flow-visibility', 'flow-aggregator')\nAND destina ...
```
Not getting l7ProtocolName? how are we checking if the flow is L7 and which protocol is used in the flow?

Dhruv-J replied:
Since only http protocols are being exported, right now I only check for httpVals existing. Once new protocols are added, their JSON will need to be parsed differently as well, so a patch PR can be made.

tushartathgur commented on plugins/grafana-custom-plugins/grafana-dependency-plugin/.config/.prettierrc.js:
```diff
@@ -0,0 +1,16 @@
+  "singleQuote": true,
+  "useTabs": false,
+  "tabWidth": 2
+};
```
Resolve all empty line issues

tushartathgur commented on plugins/grafana-custom-plugins/grafana-dependency-plugin/src/DependencyPanel.tsx:
```diff
@@ -47,124 +53,163 @@ export const DependencyPanel: React.FC<Props> = ({ options, data, width, height
+        } else {
+          graphLine = graphLine + destinationIP + ':' + destinationPort;
+        }
+        // graphLine = sourceIP + ':' + sourcePort + ' --' + frameHttpInfo.ContentLength + '--> ' + destinationIP + ':' + destinationPort + '\n';
```
remove this line?

tushartathgur commented on build/yamls/flow-visibility.yml:
```diff
@@ -3543,15 +3557,56 @@ data:
+                }
+              },
+              "queryType": "sql",
+              "rawSql": "SELECT sourcePodName, sourcePodLabels, sourcePodNamespace, sourceNodeName, destinationPodName, destinationPodLabels, destinationNodeName, destinationServicePortName, octetDeltaCount FROM flows\nWHERE sourcePodNamespace NOT IN ('kube-system', 'flow-visibility', 'flow-aggregator')\nAND destinationPodNamespace NOT IN ('kube-system', 'flow-visibility', 'flow-aggregator')\nAND ...
```
Use l7ProtocolName to filter L4 flows

tushartathgur commented on build/charts/theia/provisioning/datasources/create_table.sh:
```diff
@@ -80,6 +80,8 @@ clickhouse client -n -h 127.0.0.1 <<-EOSQL
         clusterUUID String,
         egressName String,
         egressIP String,
+        isL7 Bool,
```
l7ProtocolName?

tushartathgur commented on build/charts/theia/provisioning/dashboards/network_topology_dashboard.json:
```diff
@@ -60,11 +101,11 @@
           },
           "queryType": "sql",
-          "rawSql": "SELECT sourcePodName, sourcePodLabels, sourcePodNamespace, sourceNodeName, destinationPodName, destinationPodLabels, destinationNodeName, destinationServicePortName, octetDeltaCount FROM flows\nWHERE sourcePodNamespace NOT IN ('kube-system', 'flow-visibility', 'flow-aggregator')\nAND destinationPodNamespace NOT IN ('kube-system', 'flow-visibility', 'flow-aggregator')\nAND des ...
+          "rawSql": "SELECT sourcePodName, sourcePodLabels, sourcePodNamespace, sourceNodeName, sourceIP, sourceTransportPort, destinationPodName, destinationPodLabels, destinationNodeName, destinationServicePortName, destinationIP, destinationTransportPort, octetDeltaCount, httpVals FROM flows\nWHERE sourcePodNamespace NOT IN ('kube-system', 'flow-visibility', 'flow-aggregator')\nAND destination ...
```
Use l7ProtocolName
Refer to [Adding L7 visibility fields](https://github.com/vmware/go-ipfix/pull/315)

Dhruv-J replied:
I'll need to check httpVals regardless, since I need to parse its JSON in a specific manner, so I assumed it was redundant to check both. If needed, I can make a later patch that checks the protocolName and parses the JSON accordingly.

Dhruv-J commented:
Will further update README and changelog to reflect latest changes.

yuntanghsu commented on ci/kind/test-multi-cluster.sh:
```diff
@@ -33,7 +33,7 @@ function print_usage {
 
 TESTBED_CMD=$(dirname $0)"/kind-setup.sh"
-FLOW_VISIBILITY_CH_ONLY_CMD=$(dirname $0)"/../../hack/generate-manifest.sh --no-grafana --node-port --secure-connection"
+FLOW_VISIBILITY_CH_ONLY_CMD=$(dirname $0)"/../../hack/generate-manifest.sh --ch-only --node-port --secure-connection"
```
Remove changes not belong to this PR? Do rebase?

yuntanghsu commented on plugins/grafana-custom-plugins/grafana-dependency-plugin/README.md:
```diff
@@ -114,7 +143,8 @@ installed panels will appear. For more information, visit the docs on [Grafana p
+4 and layer 7 visualizations.
 
 <img src="https://user-images.githubusercontent.com/10016630/233191376-7cf471b8-5e3e-473a-8696-da25ab981066.png" width="400" alt="Panel Option Editor View">
 
```
I think we need to update the section 4 as you already migrate the dependency-plugin from grafana/toolkit to create-plugin?

yuntanghsu commented on .github/workflows/kind.yml:
```diff
@@ -217,6 +217,10 @@ jobs:
           docker tag antrea/theia-clickhouse-monitor:latest projects.registry.vmware.com/antrea/theia-clickhouse-monitor:latest
           docker load -i clickhouse-server.tar
           docker tag antrea/theia-clickhouse-server:latest projects.registry.vmware.com/antrea/theia-clickhouse-server:latest
+      - name: Remove Theia images tar files
```
I think we need to update the typescript.yaml in workflow as well?

Dhruv-J commented:
Unit tests are proving to be challenging to setup with updated toolkit versions, will make patch PR for them separately.

yuntanghsu commented on build/charts/theia/provisioning/datasources/migrators/000005_0-6-0.down.sql:
Do you know why we have these two empty files here?

heanlan commented on plugins/grafana-custom-plugins/grafana-dependency-plugin/src/DependencyPanel.test.tsx:
```diff
@@ -0,0 +1,45 @@
+    props.width = 600,
+    props.height = 600;
+    props.options = {};
+    let component = shallow(<DependencyPanel {...props} />);
```
@yuntanghsu could you help take a look at this UT as you recently modified chord/sankey plugins UT and have some experience with how we should design react UTs? I feel like we can also reference to your latest change https://github.com/antrea-io/theia/blob/main/plugins/grafana-custom-plugins/grafana-chord-plugin/src/ChordPanel.test.tsx, but not very sure whether it's also applicable here

yuntanghsu replied:
I think we are doing a similar test, which verifies the input data reflected in the output component.
But I think we can also verify the context for [boxColor, layerLevelFour, themeColors]? And maybe we can add one more input for layerSeven as the format of GraphString of these two layer graphs are different?

heanlan commented on plugins/grafana-custom-plugins/grafana-dependency-plugin/src/DiagramController.tsx:
```diff
@@ -0,0 +1,96 @@
+
+  renderCallback(svgCode: string, bindFunctions: any) {
+    if (this && bindFunctions) {
+      console.log('binding diagram functions');
```
debug log can be removed?

heanlan commented on plugins/grafana-custom-plugins/grafana-dependency-plugin/src/DependencyPanel.tsx:
```diff
@@ -26,12 +18,26 @@ export const DependencyPanel: React.FC<Props> = ({ options, data, width, height
   let graphString = 'graph LR;\n';
+  let styleString = '';
   let boxColor;
+  let layerLevel;
```
Why we need an extra var `layerLevel`? Can we just reuse `options.layerFour` this boolean var?

heanlan commented on plugins/grafana-custom-plugins/grafana-dependency-plugin/src/DependencyPanel.tsx:
```diff
@@ -47,124 +53,162 @@ export const DependencyPanel: React.FC<Props> = ({ options, data, width, height
+        return sourcePodName;
+      }
+  
+      let groupByPodLabel = options.groupByPodLabel;
```
same question, can we reuse `options.groupByPodLabel` without introducing new variable?

Dhruv-J replied:
yes, i can change this one

heanlan commented on plugins/grafana-custom-plugins/grafana-dependency-plugin/src/DependencyPanel.tsx:
```diff
@@ -47,124 +53,162 @@ export const DependencyPanel: React.FC<Props> = ({ options, data, width, height
+  function layerSevenGraph() {
+    function getColorFromStatus(httpStatus: string) {
+      // colors that correspond to each of the types of http response code; i.e. 4xx and 5xx codes return red to symbolize errors
+      const colors = ['orange', 'green', 'blue', 'red', 'red'];
```
I'd suggest just use a map data structure or case switch statement for this function, rather than the status first char matching colors list index.

Dhruv-J replied:
I was thinking of checking the first character because it would allow all response codes that are within the 100 possible codes range to be represented. Since all 4xx codes are error codes, and all 2xx are error codes, I just checked the first character. I can put the underlying logic into a case switch however, that would be cleaner.

heanlan replied:
Checking the first char is ok. But I think the list index thing is not a common practice.

heanlan replied:
I suppose we can combine the first char matching with a case switch statement.

heanlan reviewed (commented):
Could you add more details into the PR descriptions? including the your changes from the following aspects and why we need those changes:

1. data schema
2. dependency plugin source code changes, with newly added diagramController
3. dependency upgrade of dependency plugin, which introduces .config, jest folders, etc
4. newly added UT

heanlan commented:
Another general comment: if the PR is not ready for review, feel free to convert it to draft. If it is ready, maybe it's better to squash some intermediate and unmeaningful(to reviewers) commits in my opinion.

Dhruv-J committed:
Network Topology Dashboard Update

This PR updates the Network Topology dashboard to include visualizations for
multiple networking layers. It also updates the legends of both graphs to
improve readability.

Signed-off-by: Dhruv-J <dhruvj@vmware.com>

Dhruv-J commented on .github/workflows/typescript.yml:
```diff
@@ -64,3 +64,37 @@ jobs:
       if: always()
       run: npm uninstall -g yarn
+
+  test-unit-dependency:
```
they use different node versions, and it's not possible to specify node version individually for unit tests, so until all the plugins get updated to the same version, a separate unit test works the best

Dhruv-J committed:
addressed feedback and updated unit tests

Signed-off-by: Dhruv-J <dhruvj@vmware.com>

yuntanghsu commented on plugins/grafana-custom-plugins/grafana-dependency-plugin/package.json:
```diff
@@ -3,27 +3,72 @@
-    "start": "yarn watch"
+    "build": "webpack -c ./webpack.config.ts --env production",
+    "dev": "webpack -w -c ./webpack.config.ts --env development",
+    "e2e": "yarn exec cypress install && yarn exec grafana-e2e run",
```
Maybe you can follow this PR https://github.com/antrea-io/theia/pull/587 to remove some unused scripts and dependencies in this PR to avoid some `Vulnerability alerts` bumped up after this PR is merged.
Note:
Not all dependencies detected by [depcheck](https://www.npmjs.com/package/depcheck) is unused.

yuntanghsu commented on plugins/grafana-custom-plugins/grafana-dependency-plugin/src/DependencyPanel.test.tsx:
```diff
@@ -0,0 +1,44 @@
+      renderedHtmlString.includes(expectedGraphString)
+    ).toEqual(true);
+  });
+});
```
empty line.

yuntanghsu commented on plugins/grafana-custom-plugins/grafana-dependency-plugin/tsconfig.json:
```diff
@@ -1,9 +1,3 @@
-  }
-}
+  "extends": "./.config/tsconfig.json"
+}
```
ditto

yuntanghsu commented on plugins/grafana-custom-plugins/grafana-dependency-plugin/webpack.config.ts:
```diff
@@ -0,0 +1,20 @@
+  });
+};
+
+export default config;
```
ditto

Dhruv-J committed:
feedback fix

Signed-off-by: Dhruv-J <dhruvj@vmware.com>

Dhruv-J committed:
flow-vis yml update

Signed-off-by: Dhruv-J <dhruvj@vmware.com>

heanlan commented on plugins/grafana-custom-plugins/grafana-dependency-plugin/src/DependencyPanel.tsx:
```diff
@@ -47,124 +49,165 @@ export const DependencyPanel: React.FC<Props> = ({ options, data, width, height
+      // colors that correspond to each of the types of http response code; i.e. 4xx and 5xx codes return red to symbolize errors
+      const codeToColor = {1:'orange', 2:'green', 3:'blue', 4:'red', 5:'red'};
+      let statusType = +httpStatus.charAt(0);
+      if (statusType < 1 || statusType > 5) {
```
Is it possible for us to get statusType < 1 or > 5? Is there such a http status code? I see you added it for indicating errors, but I wonder whether it is possible to happen, if it does happen, would it be better to catch this error in Flow Exporter or Flow Aggregator? cc @tushartathgur

Dhruv-J replied:
I don't think it should, but we can double check with @tushartathgur.

heanlan commented on plugins/grafana-custom-plugins/grafana-dependency-plugin/src/DiagramController.tsx:
```diff
@@ -0,0 +1,97 @@
+            ref={this.setDiagramRef}
+            className={`diagram-seven`}
+          ></div>
+          <div><p>In this graph, each flow between a source and destination is represented by an arrow labelled with the http content length of the data being sent.
```
Do we need a legend/sentence to explain how the colors are mapped to http codes, e.g. 1xx(informational response), 2xx(success), etc?

Dhruv-J replied:
Yes, I can add a sentence below at the end.

Dhruv-J committed:
http codes included in legend

Signed-off-by: Dhruv-J <dhruvj@vmware.com>

Dhruv-J commented:
Instructions for those wanting to deploy the solution for themselves:

1. pull Antrea PR [#5218](https://github.com/antrea-io/antrea/pull/5218)
2. pull Theia PR #417 and this PR
3. execute the included script for [build_antrea.sh.zip](https://github.com/antrea-io/theia/files/13639064/build_antrea.sh.zip)
4. deploy flow-visibility.yml from the new theia PR
5. apply the following: [pods.yaml.zip](https://github.com/antrea-io/theia/files/13639121/pods.yaml.zip)
6. annotate the pods with the following command: `kubectl pod annotate <pod-name> -n <namespace> [visibility.antrea.io/enable-l7=both](http://visibility.antrea.io/enable-l7=both)`
7. kubectl exec into the applied admin pod and run the command `curl example.com` to get one row, or `curl example.com example.com example.com` to get multiple rows (it'll take about 2 minutes to populate clickhouse)
8. open the application layer dashboard to view http values

heanlan reviewed (commented):
LGTM

Let's wait for all the required PRs getting merged first and give the final approval on this
